### ISSUE TITLE
Race condition happens in HttpRequest if you modify reference count in two threads

### ISSUE THREAD

**Issue by tanwenzhao** [cocos2d/cocos2d-x#19169]
- cocos2d-x version: 3.* and 2.*
This is a serious issue and exists in both 2.* and 3.*,
Code below is a typical code of HttpRequest in version 2.*, there is race condition btweeen "request->release()" and reference count changes to request in networkThread funciton at HttpClient.cpp. This will cause request has wrong reference count and sometimes be freed and used later, causing crash and random behavior. If you use userdata in request, passing in function pointers or write to buffer, this will cause crash.

In version 3.*, you still change request's reference count in networkThread when dequing the request from s_requestQueue. So there is also possible race conditions if the upper caller code is similar to below with request->release()

Our temporary fix is to remove both the request->release() in main thread and the retain in send().
Race condition happens in HttpRequest if you modify reference count in two threads. This is a general potential problem and need to pay lots of attention.

HttpRequest* request = new HttpRequest();
request->setRequestType(HttpRequest::Type::GET);
request->setUrl("http://YOUR_URL");
request->setResponseCallback(CC_CALLBACK_2(HttpClientTest::onHttpRequestCompleted, this));
request->setTag("Get test");
HttpClient::getInstance()->send(request);
request->release();

**Comment by zcgit** [comment]
yes, you should use HttpClient::getInstance()->sendImmediate(req), hope this is useful for you :)

**Comment by tanwenzhao** [comment]
you are right. I briefly took a look of sendImmediate, it does not change request's reference count in worker thread.
